cmd/gardener-admission-controller/app/gardener_admission_controller.go

```diff
@@ -0,0 +1,221 @@
+// decodeConfig decodes data as a ControllerManagerConfiguration object.
+func (o *options) decodeConfig(data []byte) (*config.AdmissionControllerConfiguration, error) {
+	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	config, ok := configObj.(*config.AdmissionControllerConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("got unexpected config type: %v", gvk)
+	}
+	return config, nil
+}
+
+func (o *options) applyDefaults(in *config.AdmissionControllerConfiguration) (*config.AdmissionControllerConfiguration, error) {
+	external, err := o.scheme.ConvertToVersion(in, admissioncontrollerconfigv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	o.scheme.Default(external)
+
+	internal, err := o.scheme.ConvertToVersion(external, config.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := internal.(*config.AdmissionControllerConfiguration)
+
+	return out, nil
+}
+
+// AdmissionController contains all necessary information to run the admission controller.
+type AdmissionController struct {
+	Config    *config.AdmissionControllerConfiguration
+	ClientMap clientmap.ClientMap
+}
+
+func (o *options) run(ctx context.Context) error {
+	c, err := o.loadConfigFromFile(o.ConfigFile)
+	if err != nil {
+		return err
+	}
+
+	if c, err = o.applyDefaults(c); err != nil {
+		return err
+	}
+
+	admissionController, err := NewAdmissionController(c)
+	if err != nil {
+		return err
+	}
+
+	return admissionController.Run(ctx)
+}
+
+// NewAdmissionController creates a new AdmissionController instance.
+func NewAdmissionController(cfg *config.AdmissionControllerConfiguration) (*AdmissionController, error) {
+	if cfg == nil {
+		return nil, errors.New("config is required")
+	}
+
+	// Initialize logger
+	logger := logger.NewLogger(cfg.LogLevel)
+	logger.Info("Starting Gardener admission controller...")
+
+	// Prepare a Kubernetes client object for the Garden cluster which contains all the Clientsets
+	// that can be used to access the Kubernetes API.
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.GardenClientConnection.Kubeconfig = kubeconfig
+	}
+
+	restCfg, err := kubernetes.RESTConfigFromClientConnectionConfiguration(&cfg.GardenClientConnection, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Enable cache globally for garden client.
+	kubernetes.UseCachedRuntimeClients = true
+
+	clientMap, err := clientmapbuilder.NewGardenClientMapBuilder().WithLogger(logger).WithRESTConfig(restCfg).Build()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to build ClientMap: %w", err)
+	}
+
+	return &AdmissionController{
+		ClientMap: clientMap,
+		Config:    cfg,
+	}, nil
+}
+
+// Run starts the Gardener admission controller.
+func (a *AdmissionController) Run(ctx context.Context) error {
+	k8sGardenClient, err := a.ClientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return err
+	}
+
+	if err := a.ClientMap.Start(ctx.Done()); err != nil {
```

Comment:
I think, one problem that we might hit here, is that we now don't wait for the needed informers to sync before we start the webhook server.
One option would be to retrieve an informer for every resource we want to talk to and then use the `HasSynced` funcs to wait for the caches to sync.

WDYT about constructing the webhook handlers before starting the server, then passing the client to some kind of `Initialize` func there and let the handlers retrieve the needed informers and necessary `HasSynced` funcs?

Something like this would help to make sure, the webhook can directly serve requests off the cache, when it starts accepting traffic, and not to introduce to much unneeded latency, when calling the gardener api.

Reply:
I was a bit more pragmatic and hope you are fine with the changes.

Reply:
Hmm, ok. I would have to test that, but I would definitely expect some latency on startup in larger landscapes because the handler has to load all Shoots and all Projects. So, I am still a bit concerned about this.

Another option to mitigate this situation and still keep it pragmatic in code, would be to add a `startupProbe` (or `livenessProbe` with some `initialDelaySeconds`).
Still, this would only be based on heuristics and not really detect any long running initial syncs...
/cc @rfranzke

Reply:
Sorry, I have just now seen that you added the `HasSynced` funcs, that's great :)